notebooks/403-action-recognition-webcam/403-action-recognition-webcam.ipynb

```diff
@@ -0,0 +1,646 @@
+    "    \n",
+    "    x: np.array, axis=None\n",
+    "    \"\"\"\n",
+    "    exp = np.exp(x)\n",
+    "    return exp / np.sum(exp, axis=axis)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    },
+    "tags": []
+   },
+   "source": [
+    "### Main Processing Function\n",
+    "\n",
+    "Run action recognition function will run in different operations, either a webcam or a video file. See the list of procedures below:\n",
+    "\n",
+    "1. Create a video player to play with target fps (`utils.VideoPlayer`).\n",
+    "2. Prepare a set of frames to be encoded-decoded.\n",
+    "3. Run AI functions\n",
+    "4. Visualize the results."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def run_action_recognition(source: str = 0, flip: bool = False, use_popup: bool = False, skip_first_frames: int = 0):\n",
+    "    \"\"\"\n",
+    "    Use the \"source\" webcam or video file to run the complete pipeline for action-recognition problem\n",
+    "    1. Create a video player to play with target fps\n",
+    "    2. Prepare a set of frames to be encoded-decoded\n",
+    "    3. Preprocess frame before Encoder\n",
+    "    4. Encoder Inference per frame\n",
+    "    5. Decoder inference per set of frames\n",
+    "    6. Visualize the results\n",
+    "    \n",
+    "    :param: source: webcam \"0\" or video path\n",
+    "    :param: flip: to be used by VideoPlayer function for flipping capture image\n",
+    "    :param: use_popup: False for showing encoded frames over this notebook, True for creating a popup window.\n",
+    "    :param: skip_first_frames: to be used by VideoPlayer function for skipping first frames\n",
+    "    :returns: display video over the notebook or in a popup window\n",
+    "    \n",
+    "    \"\"\"\n",
+    "    \n",
+    "    sample_duration=16\n",
+    "    size=224\n",
+    "    #create a video player to play with target fps\n",
+    "    player = utils.VideoPlayer(source, flip=flip, fps=30, skip_first_frames=skip_first_frames)\n",
+    "    # start capturing\n",
+    "    player.start()\n",
+    "    try:\n",
+    "        if use_popup:\n",
+    "            title = \"Press ESC to Exit\"\n",
+    "            cv2.namedWindow(title, cv2.WINDOW_GUI_NORMAL | cv2.WINDOW_AUTOSIZE)\n",
+    "        \n",
+    "        processing_times = collections.deque()\n",
+    "        frames = []\n",
+    "        video_frames = []\n",
+    "        encoder_output = []\n",
+    "        decoded_labels = [0, 0, 0]\n",
+    "        decoded_top_probs = [0, 0, 0]\n",
+    "        while True:\n",
+    "            \n",
+    "            encoder_output = []\n",
+    "            # loop over the number of required sample frames\n",
+    "            # Prepare a set of frames to be enconded-decoded\n",
+    "            for i in range(0, sample_duration):\n",
```

Comment:
The video is limited to 30 fps above, which is good for display, but that means we have to wait for half a second before all the frames are collected and sent to the model. That is unfortunate.

Reply:
Yes, We will have a delay of 0.5 seconds. I was working on an Async solution, but that is possibly another notebook.

Reply:
You can also update this one if you have a nice solution later. We have async inference in 110, but that is maybe overkill for this notebook, because the model is fast enough.

Reply:
This could be done without the async pipeline. Just use a sliding window. You'll have new results every frame.